### ISSUE TITLE
Cast to date fails if date is in DD-MM-YY format

### ISSUE THREAD

**Issue by Anish9901** [mathesar-foundation/mathesar#3953]
## Description

<img width="1486" alt="Screenshot 2024-10-11 at 8 12 28 PM" src="https://github.com/user-attachments/assets/907bf4c8-20a8-4b55-ac56-30aaa2adeacb">

## Expected behavior

Casting shouldn't fail

## To Reproduce

1. Create a table from scratch
2. Create a `Text` column named "Date"
3. Add `22/06/24` as a record in the date column.
4. Try casting the Date column to `DATE` type.

**Comment by Anish9901** [comment]
No, this is not a regression, we can set different datestyles on postgres at runtime but there is no way to do it via the UI at the moment.

**Comment by seancolsen** [comment]
@Anish9901 if a text column contained one cell with the string "01/02/03", how would you expect Mathesar to cast it? Would it become January 2nd or February 1st? Or would the user (or potentially Mathesar administrator) be granted some control over this behavior? While much of the world might expect the date to be in February, US users would likely expect it to be January.

**Comment by mathemancer** [comment]
1. I don't think this should be in the beta milestone.
2. We're likely to lose functionality if we're not careful about modifying this behavior, since the Postgres date casting behavior is pretty much as good as possible, subject to the issue that @seancolsen raised w.r.t. locale differences. For example, the following are all correctly handled:
  - `mathesar=# SELECT '2001-12-15'::date;`
  - `mathesar=# SELECT '2001/12/15'::date;`
  - `mathesar=# SELECT '20011215'::date;`
  - `mathesar=# SELECT '12-15-2001'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT '12/15/2001'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT '12/1/2001'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT '1/1/2001'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT '1.1.2001'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT '12/15/2001 BC'::date; -- U.S. locale assumed`
  - `mathesar=# SELECT 'J2452259'::date; -- Julian date`
  - `mathesar=# SELECT '0034-12-15 BC'::date;`
  - `mathesar=# SELECT 'December 15 2001 BC'::date;`
  - `mathesar=# SELECT 'Tuesday, October 29, 2024'::date;`
  - `mathesar=# SELECT 'Tuesday, October 29, 2024 8:42 AM'::date;`

I could go on, but I assume the point is made. We are simply not going to do better than that with home-rolled logic. Moreover, I don't think it makes any sense to give up that power or flexibility by making the user choose a format that has to apply to an entire column (which could be from an inconsistent source), and certainly not with a picker that's likely to lack many (or most) of the options available out-of-the-box in Postgres.

With all that said, the problem is real. I suggest a dropdown that simply allows the user to choose the DateStyle parameter if desired (with some helpful hints). See [the docs](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-INPUT) for what that would look like. The short version is that the parameter lets you specify how Postgres should choose between indeterminate options for, e.g., `mathesar=# SELECT '1/1/24'::date;` without constraining the behavior to a specific format.

It's possible to make this setting on a per-session or per-transaction basis. I suggest per-transaction, with the setting submitted as an optional parameter when calling the cast function. We shouldn't change the response format, since that's already handled properly.

**Comment by radheshpai87** [comment]
Hey @Anish9901 i suppose this is the issue you were talking about can we work on it. Maybe try and raise a PR for it.
